ARCHITECTURE.md

```diff
@@ -0,0 +1,275 @@
+interpreted as a key/value pair.
+
+Now Bob writes a value to his feed, and then Alice and Bob sync. The result is:
+
+```
+// Feed
+
+//// Alice
+0 (/foo/bar = 'baz')
+1 (/foo/2   = '{ "some": "json" }')
+2 (''       = '')
+
+//// Bob
+0 (/a/b     = '12')
+
+
+// Graph
+
+Alice: 0  <---  1  <---  2
+Bob  : 0
+```
+
+Notice that none of Alice's entries refer to Bob's, and vice versa. This is
+because neither has written any entries to their feeds since the two became
+aware of each other (authorized & replicated each other's feeds).
+
+Next, Alice writes a new value, and her latest entry will refer to Bob's:
+
+```
+// Feed
+
+//// Alice
+0 (/foo/bar = 'baz')
+1 (/foo/2   = '{ "some": "json" }')
+2 (''       = '')
+3 (/foo/hup = 'beep')
+
+//// Bob
+0 (/a/b     = '12')
+
+
+// Graph
+
+Alice: 0  <---  1  <---  2  <--/  3
+Bob  : 0  <-------------------/
+```
+
+Because Alice's latest feed entry refers to Bob's latest feed entry, there is
+now only one "head" in the database. That means there is enough information in
+Alice's seq=3 entry to find any other key in the database. In the last example,
+there were two heads (Alice's seq=2 and Bob's seq=0); both of which would need
+to be read internally in order to locate any key in the database.
+
+## Authorization
+
+The set of hypercores are *authorized* in that the original author of the first
+hypercore in a hyperdb must explicitly denote in their append-only log that the
+public key of a new hypercore is permitted to edit the database. Any authorized
+member may authorize more members. There is no revocation or other author
+management elements currently.
+
+## Incremental index
+
+HyperDB builds an *incremental index* with every new key/value pairs ("nodes")
+written. This means a separate data structure doesn't need to be maintained
+elsewhere for fast writes and lookups: each node written has enough information
+to look up any other key quickly and otherwise navigate the database.
+
+Each node stores the following basic information:
+
+- `key`: the key that is being created or modified. e.g. `/home/sww/dev.md`
+- `value`: the value stored at that key.
+- `seq`: the sequence number of this entry in the owner's hypercore. 0 is the
+  first, 1 the second, and so forth.
+- `feed`: the ID of the hypercore writer that wrote this
+- `path`: a 2-bit hash sequence of the key's components
+- `trie`: a navigation structure used with `path` to find a desired key
+- `clock`: vector clock to determine node insertion causality
+- `feeds`: an array of { feedKey, seq } for decoding a `clock`
+
+### Vector clock
+
+Each node stores a [vector clock](https://en.wikipedia.org/wiki/Vector_clock) of
+the last known sequence number from each feed it knows about. This is what forms
+the DAG structure.
+
+For example, Alice's vector clock on her seq=3 entry above would be `[3, 0]`
+since she knows of her latest entry (seq=3) and Bob's (seq=0).
+
+*TODO: is this correct? ^^^*
```

Comment:
Does this mean that the vector clock’s dimensions increase with the number of authorised writers?

Reply:
Exactly! :tada: